rasa/utils/tensorflow/rasa_layers.py

```diff
@@ -0,0 +1,815 @@
+            training: Python boolean indicating whether the layer should behave in
+                training mode (applying dropout to sparse tensors if applicable) or in
+                inference mode (not applying dropout).
+
+        Returns:
+            combined features: a 3-D tensor containing the combined sequence- and sentence-
+                level, sparse & dense features.
+            mask_combined_sequence_sentence: a binary 3-D tensor with 1s in place of real
+                features in the combined feature array, and 0s in place of fake features.
+        """
+        sequence_features = inputs[0]
+        sentence_features = inputs[1]
+        sequence_feature_lengths = inputs[2]
+
+        # This mask is specifically for sequence-level features.
+        mask_sequence = _compute_mask(sequence_feature_lengths)
+
+        # Process sequence-level features if any are present.
+        if self._feature_types_present[SEQUENCE]:
+            sequence_features_combined = self._tf_layers[f"sparse_dense.{SEQUENCE}"](
+                (sequence_features,), training=training
+            )
+
+            # apply mask which has 1s at positions of real tokens and 0s at all padded
+            # token positions. This is needed because the sparse+dense combining layer
+            # might've turned some fake (padded) features (i.e. 0s) into non-zero numbers
+            # and we want those to become zeros again.
+            # This step isn't needed for sentence-level features because those are never
+            # padded -- the effective sequence length in their case is always 1.
+            sequence_features_combined = sequence_features_combined * mask_sequence
+
+        # Process sentence-level features if any are present.
+        if self._feature_types_present[SENTENCE]:
+            sentence_features_combined = self._tf_layers[f"sparse_dense.{SENTENCE}"](
+                (sentence_features,), training=training
+            )
+            # Sentence-level features have sequence dimension of length 1, add it to
+            # sequence-level feature lengths.
+            combined_sequence_sentence_feature_lengths = sequence_feature_lengths + 1
+        else:
+            # Without sentence-level features, the feature lengths are completely
+            # determined by sequence-level features.
+            combined_sequence_sentence_feature_lengths = sequence_feature_lengths
+
+        mask_combined_sequence_sentence = _compute_mask(
+            combined_sequence_sentence_feature_lengths
+        )
+
+        # If both feature types are present, combine them. Otherwise just the present
+        # feature type will be returned.
+        if (
+            self._feature_types_present[SEQUENCE]
+            and self._feature_types_present[SENTENCE]
+        ):
+            features_to_return = self._concat_sequence_sentence_features(
+                sequence_features_combined,
+                sentence_features_combined,
```

Comment:
pycharm complains that these two might not exist because they are defined inside the `if`

Reply:
Well... I apologise to PyCharm but from those `if`s it's clear that they will exist exactly when needed. Would you change it in the code? If yes, how?

Reply:
Never mind, it's now addressed.